### PR TITLE
For super admin, add pagination on `/admin/enterprises`

### LINKED ISSUE #8027: Performance improvement on /admin/enterprises
For context see https://github.com/openfoodfoundation/openfoodnetwork/issues/8020.

### PR DESCRIPTION
**Pull request by jibees**
#### What? Why?

Introducing pagination thanks to `pagy` for super admin on this page.

Also, previously, we used to use `@enterprise_set = Sets::EnterpriseSet.new(collection)` instead of creating set based on `@collection`: this leads to call the `collection` method twice, which was probably very time consuming. This commit fix also that.

- Closes #8027

#### What should we test?

- As a super admin, we should this a faster page and paginated on `/admin/enterprises`. In term of performance, actually, I don't know if this could be mesurable in staging. Maybe we could compare the request time of `/admin/enterprises` URL via chrome|firefox inspector. 
- As a hub manager, nothing should change.

#### Release notes

Changelog Category: User facing changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: app/controllers/admin/enterprises_controller.rb [hunk 1]
```diff
@@ -8,12 +8,13 @@ module Admin
...
     include GeocodeEnterpriseAddress
     include CablecarResponses
+    include Pagy::Backend
 
     # These need to run before #load_resource so that @object is initialised with sanitised values
     prepend_before_action :override_owner, only: :create
     prepend_before_action :override_sells, only: :create
 
-    before_action :load_enterprise_set, only: :index
+    before_action :load_enterprise_set_on_index, only: :index
```
Comment (mkllnk) [inline app/controllers/admin/enterprises_controller.rb:17]: When something is happening only in one action, I often prefer to call it within that action. It keeps related code together and you can see quicker what's happening.

**Review by mkllnk — APPROVED** [review]
Nice quick fix!

File: app/controllers/admin/enterprises_controller.rb (hunk 1 above)
Comment (jibees) [reply app/controllers/admin/enterprises_controller.rb]: I agree 👍 
84e2fbc60a

File: app/controllers/admin/enterprises_controller.rb [hunk 2]
```diff
@@ -148,8 +150,15 @@ def find_resource
...
-    def load_enterprise_set
-      @enterprise_set = Sets::EnterpriseSet.new(collection) if spree_current_user.admin?
+    def load_enterprise_set_on_index
+      return unless spree_current_user.admin?
+
+      load_enterprise_set_with_params
+    end
+
+    def load_enterprise_set_with_params(params = {})
+      @pagy, paginated_collection = pagy(@collection)
```
Comment (dacook) [inline app/controllers/admin/enterprises_controller.rb:161]: I'm curious, why the extra local variable instead of just using the instance variable?

**Review by dacook — APPROVED** [review]
Great, I can't wait to see this in action!

File: app/controllers/admin/enterprises_controller.rb (hunk 2 above)
Comment (jibees) [reply app/controllers/admin/enterprises_controller.rb]: You mean, something like:

```ruby
@pagy, @paginated_collection = pagy(@collection)
```
?

The only reason (which is not actually) I have is that it's done like this in the other part of the code.
But you're right, [documentation](https://github.com/ddnexus/pagy) is clear about this:
```ruby
# Wrap your collections with pagy in your actions
@pagy, @records = pagy(Product.all)
```

Conclusion: done! (and force pushed)

File: app/controllers/admin/enterprises_controller.rb [hunk 3]
```diff
@@ -148,8 +151,15 @@ def find_resource
...
-      @enterprise_set = Sets::EnterpriseSet.new(collection) if spree_current_user.admin?
+    def load_enterprise_set_on_index
+      return unless spree_current_user.admin?
+
+      load_enterprise_set_with_params
+    end
+
+    def load_enterprise_set_with_params(params = {})
+      @pagy, @paginated_collection = pagy(@collection)
+      @enterprise_set = Sets::EnterpriseSet.new(@paginated_collection, params)
```
Comment (mkllnk) [inline app/controllers/admin/enterprises_controller.rb:162]: I guess that `@paginated_collection` is not used anywhere else in the code. So by making it a local variable, it's clear that you can change the name or do whatever you like without affected other code.

On the other hand, this variable may be useful at some point. And if we publicise `@pagy` then why not the collection? So I don't mind this version either. It looks more consistent.

**Review by mkllnk — APPROVED** [review]

File: app/controllers/admin/enterprises_controller.rb (hunk 2 above)
Comment (dacook) [reply app/controllers/admin/enterprises_controller.rb]: Apologies for the confusion, I actually didn't understand the code and assumed that the same object was being stored in both variables. 
But I guess the result of the `pagy` method is an array with two elements, assigned to each variable. 

Good idea to use the convention from the docs 👍

**Pull request merged**